### PR TITLE
Drawing engine

### PR DESCRIPTION
**Pull request by kecajjo**

### CODE REVIEW

**Review by ProrokWielki — APPROVED** [review]

**Review by ProrokWielki — APPROVED** [review]

**Comment by ProrokWielki** [comment]
Could you split this PR into two? one for clang stuff and the second one for implementation

**Comment by kecajjo** [comment]
> Could you split this PR into two? one for clang stuff and the second one for implementation

Ill try after i fix all clang tidy stuff

**Review by kecajjo — COMMENTED** [review]
CI will pass after clang tidy fixes are merged

File: CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,7 @@ if(${TIDY})
 endif()
 
 add_subdirectory(bl)
+add_subdirectory(gui)
 
 add_executable(tetris main/main.cpp)
-target_link_libraries(tetris PUBLIC blocks board)
+target_link_libraries(tetris PUBLIC blocks board frame_buffer)
```
Comment (ProrokWielki) [inline CMakeLists.txt]: I'd say frame_buffer should be private dependency of the GUI lib (or what ever you want to call it), There is no need for the fianal user to even know there is such thing as frame buffer. The user should have only acces to drawing engine(interface and factory), abstract drawable and Color

File: gui/drawable/inc/drawable.hpp [hunk 2]
```diff
@@ -29,17 +27,18 @@ class Drawable {
     const std::unordered_map<std::shared_ptr<Drawable>, Drawable::Position> &getChildDrawables()
```
Comment (ProrokWielki) [inline gui/drawable/inc/drawable.hpp:26]: shoudl be const

File: gui/drawable/inc/drawable.hpp [hunk 3]
```diff
@@ -29,17 +27,18 @@ class Drawable {
...
+    void setColor(const Color &c) { color = c; }
 
    protected:
     explicit Drawable() = default;
 
-    // TODO change in the drawing engine PR
-    // NOLINTBEGIN
-    Color color = {max_color_val, max_color_val, max_color_val, max_color_val};
+    // NOLINTNEXTLINE
     std::list<Position> shape;
```
Comment (ProrokWielki) [inline gui/drawable/inc/drawable.hpp:37]: what is wrong with this line - why does it need NOLINTNEXTLINE

File: gui/drawable/inc/drawable.hpp [hunk 4]
```diff
@@ -29,17 +27,18 @@ class Drawable {
...
-    std::unordered_map<std::shared_ptr<Drawable>, Position> children;
-    // NOLINTEND
 
    private:
+    std::unordered_map<std::shared_ptr<Drawable>, Position> children;
+    Color color = {max_color_val, max_color_val, max_color_val, max_color_val};
     static constexpr unsigned max_color_val = std::numeric_limits<unsigned char>::max();
```
Comment (ProrokWielki) [inline gui/drawable/inc/drawable.hpp]: this should be in Color struct

**Review by ProrokWielki — CHANGES_REQUESTED** [review]

File: gui/drawing_engine/src/drawing_engine.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    return getPixel(p.x, p.y);
+}
+
+void DrawingEngine::setPixel(const Color& c, const unsigned int& x, const unsigned int& y) {
+    fb.at(currentFb).setPixel(x, y, c);
+}
+
+void DrawingEngine::setPixel(const Color& c, const Drawable::Position& p) { setPixel(c, p.x, p.y); }
+
+void DrawingEngine::drawDrawable(const Drawable* const drawable, const Drawable::Position& p) {
```
Comment (ProrokWielki) [inline gui/drawing_engine/src/drawing_engine.cpp]: take smart pointer or reference here

**Review by ProrokWielki — CHANGES_REQUESTED** [review]

File: gui/drawable/inc/drawable.hpp (hunk 3 above)
Comment (kecajjo) [reply gui/drawable/inc/drawable.hpp:37]: It is a protected variable inside a class, only private members are allowed

Comment (ProrokWielki) [reply gui/drawable/inc/drawable.hpp:37]: wait, clang tidy does not allow for protected variables? what is the exact rule?

Comment (kecajjo) [reply gui/drawable/inc/drawable.hpp:37]: [cppcoreguidelines-non-private-member-variables-in-classes](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/non-private-member-variables-in-classes.html)

File: gui/drawable/inc/drawable.hpp (hunk 2 above)
Comment (kecajjo) [reply gui/drawable/inc/drawable.hpp:26]: it is const, but clang-format moved "const" word to the next line

File: gui/drawable/inc/drawable.hpp (hunk 4 above)
Comment (kecajjo) [reply gui/drawable/inc/drawable.hpp]: fixed

File: CMakeLists.txt (hunk 1 above)
Comment (kecajjo) [reply CMakeLists.txt]: fixed

File: gui/drawing_engine/src/drawing_engine.cpp (hunk 5 above)
Comment (kecajjo) [reply gui/drawing_engine/src/drawing_engine.cpp]: fixed

**Review by ProrokWielki — APPROVED** [review]

**Pull request merged**